### PR TITLE
Add octospim support for octospi

### PR DESCRIPTION
**Pull request by HaoboGu**
This PR adds initial support for OCTOSPIM peripheral of octospi. Current implementation just adds OCTOSPIM to octospi, and enables corresponding pins of OCTOSPIM. 

Closes https://github.com/embassy-rs/embassy/issues/3097

### CODE REVIEW

**Comment by HaoboGu** [comment]
I've tested this PR on stm32h7b0 + w25q64 qspi flash, worked well. But there are still limitations of current implementation:

1. Current implementation supports OCTOSPI1 only. Different octospim register value should be configured for OCTOSPI1/2, but `Ospi` struct has no information about it.
2. In `build.rs`, I also hardcoded `OCTOSPI1` for `pin_trait_impl`.

**Comment by HaoboGu** [comment]
Did more testing on a rm67162 qspi screen, worked too.

**Comment by ninjasource** [comment]
Hi @HaoboGu, I managed to use your branch on an stm32h735g-dk to communicate with PSRAM through the hyperbus. However, I had to use OCTOSPI2 because of how the dev kit is wired up so I think it's worth trying to figure out how to support OCTOSPI2 as well. Are you still working on this branch?

**Comment by HaoboGu** [comment]
@ninjasource Yeah, OCTOSPI2 is not support now. I had some problems when implementing it for both OCTOSPI1 and OCTOSPI2. For a temporary solution, you can change [here](https://github.com/embassy-rs/embassy/pull/3102/files#diff-c33bc057fbbc72f92108a966f9a86862be7e612e1abfb8f43804190e38efc6efR1095) and [here](https://github.com/embassy-rs/embassy/pull/3102/files#diff-3afb3f71bbd57a91adea9e4a639e1101c3a2d74a972104515fe3531e3edcbdf4R246-R247) to make it work for OCTOSPI2

**Comment by HaoboGu** [comment]
I've pushed a commit making it possible to support both OCTOSPI1 and OCTOSPI2. But 

> Different octospim register value should be configured for OCTOSPI1/2, but Ospi struct has no information about it.

This problem is still there. @Dirbaio Do you have any suggestions on solving it?

File: embassy-stm32/build.rs [hunk 1]
```diff
@@ -1040,6 +1040,18 @@ fn main() {
...
+        (("octospim", "P1_IO3"), quote!(crate::ospi::D3Pin)),
+        (("octospim", "P1_IO4"), quote!(crate::ospi::D4Pin)),
+        (("octospim", "P1_IO5"), quote!(crate::ospi::D5Pin)),
+        (("octospim", "P1_IO6"), quote!(crate::ospi::D6Pin)),
+        (("octospim", "P1_IO7"), quote!(crate::ospi::D7Pin)),
+        (("octospim", "P1_DQS"), quote!(crate::ospi::DQSPin)),
+        (("octospim", "P1_NCS"), quote!(crate::ospi::NSSPin)),
+        (("octospim", "P1_CLK"), quote!(crate::ospi::SckPin)),
+        (("octospim", "P1_NCLK"), quote!(crate::ospi::NckPin)),
         (("tsc", "G1_IO1"), quote!(crate::tsc::G1IO1Pin)),
```
Comment (ninjasource) [inline embassy-stm32/build.rs:1055]: Some missing pins:
```rust
        (("octospim", "P1_IO0"), quote!(crate::ospi::D0Pin)),
        (("octospim", "P1_IO1"), quote!(crate::ospi::D1Pin)),
        (("octospim", "P1_IO2"), quote!(crate::ospi::D2Pin)),
        (("octospim", "P1_IO3"), quote!(crate::ospi::D3Pin)),
        (("octospim", "P1_IO4"), quote!(crate::ospi::D4Pin)),
        (("octospim", "P1_IO5"), quote!(crate::ospi::D5Pin)),
        (("octospim", "P1_IO6"), quote!(crate::ospi::D6Pin)),
        (("octospim", "P1_IO7"), quote!(crate::ospi::D7Pin)),
        (("octospim", "P1_DQS"), quote!(crate::ospi::DQSPin)),
        (("octospim", "P1_NCS"), quote!(crate::ospi::NSSPin)),
        (("octospim", "P1_CLK"), quote!(crate::ospi::SckPin)),
        (("octospim", "P1_NCLK"), quote!(crate::ospi::NckPin)),        
        (("octospim", "P2_IO0"), quote!(crate::ospi::D0Pin)),
        (("octospim", "P2_IO1"), quote!(crate::ospi::D1Pin)),
        (("octospim", "P2_IO2"), quote!(crate::ospi::D2Pin)),
        (("octospim", "P2_IO3"), quote!(crate::ospi::D3Pin)),
        (("octospim", "P2_IO4"), quote!(crate::ospi::D4Pin)),
        (("octospim", "P2_IO5"), quote!(crate::ospi::D5Pin)),
        (("octospim", "P2_IO6"), quote!(crate::ospi::D6Pin)),
        (("octospim", "P2_IO7"), quote!(crate::ospi::D7Pin)),
        (("octospim", "P2_DQS"), quote!(crate::ospi::DQSPin)),
        (("octospim", "P2_NCS"), quote!(crate::ospi::NSSPin)),
        (("octospim", "P2_CLK"), quote!(crate::ospi::SckPin)),
        (("octospim", "P2_NCLK"), quote!(crate::ospi::NckPin)),
```

Comment (HaoboGu) [reply embassy-stm32/build.rs:1067]: Ah, I missed it! Will add it soon

**Comment by HaoboGu** [comment]
Added a way to support both OCTOSPI1 and OCTOSPI2

**Comment by HaoboGu** [comment]
This PR can be merged now.

**Review by Dirbaio — APPROVED** [review]
thank you!

**Comment by ninjasource** [comment]
Hi @HaoboGu, thanks for your work on octospim. However, I cannot find any examples I can test and your eg-demo linked above is a broken link. Do you have any examples for us to look at?

**Comment by HaoboGu** [comment]
> Hi @HaoboGu, thanks for your work on octospim. However, I cannot find any examples I can test and your eg-demo linked above is a broken link. Do you have any examples for us to look at?

Hey @ninjasource I just made the project public

**Comment by ninjasource** [comment]
> > Hi @HaoboGu, thanks for your work on octospim. However, I cannot find any examples I can test and your eg-demo linked above is a broken link. Do you have any examples for us to look at?
> 
> Hey @ninjasource I just made the project public

Thanks! 

I have raised an issue I encountered in your repo so as not to add noise to this closed repo.

**Pull request merged**